### PR TITLE
Different focus selection for different widths

### PR DESCRIPTION
**Pull request by geetanshu-m**
### Summary of Changes

Now, We can define different focus area for different display width (optionally). 

- If in case the user has not defined the focus for a particular width, then the maximum of all the focus areas has been chosen and shown.

- In case, User only defines the focus area for one width only, It would be considered as the focus area of all the possible width.

### Testing Instructions

- In media-manager upload any image.
- Go to edit option of that image, goto Smart Crop tab.
-  Select a focus area for the given width.
-  Select another focus area by 
	- Changeing the width value from the combo box
	- Then selecting the focus for that width.
- After that Save and Close the edit tab.
- Now add that image to the article and see If the image changes when the browser window width is changed.

Note : Please check if following plugins are enabled or not (If not please enable them)
1. adaptiveimage (Content Plugin)
2. smartcrop (Media-Action Plugin)

### Expected result

The changes that the user may observe.
1. The images changes at the front-end when the width of the view port (window size) is being changed.
2. The cropped images can be found in ```/images/.cache```.

#### Selecting the focus area for diffrent width of the view port.
screenshot 16
screenshot 17
screenshot 18
screenshot 19
#### Cropped images for diffrent width
screenshot 20

### CODE REVIEW

File: media/plg_media-action_smartcrop/css/smartcrop.css [hunk 1]
```diff
@@ -1 +1,5 @@
-/* This is the CSS File for Smart Crop */
\ No newline at end of file
+/* This is the CSS File for Smart Crop */
+#attrib-smartcrop
+{
+    overflow-y: auto
+}
```
Comment (kasvith) [inline media/plg_media-action_smartcrop/css/smartcrop.css]: new line missing

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 2]
```diff
@@ -61,80 +157,18 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
...
-	function resolveBaseUrl() {
-		var basePath = window.location.origin;
-		var url = window.location.pathname.split('/');
-		if (url[1]!='administrator') {
-			return basePath+"/"+url[1];
-		}
-		return basePath;
-	}
-
 })();
```
Comment (kasvith) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: new line missing

**Review by kasvith — COMMENTED** [review]
new lines are missing

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 3]
```diff
@@ -8,7 +8,107 @@ Joomla.MediaManager = Joomla.MediaManager || {};
...
+
+    // Setting the focus area in the editor
+    var getFocusPoints = function(width){
+        var data, path, 
+            xhr, url;
+        // resolveing the path
+        path = getQueryVariable('path').split(':');
+        path = '/images' + path[1];
+
+        xhr = new XMLHttpRequest();
```
Comment (anuragteapot) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: Can we use <code>Joomla.request</code> ? 
Like this https://github.com/joomla-projects/gsoc18_adaptive_images/blob/87dcea468b911f405c46742be3a145de2c4ddcad/media/plg_quickicon_extensionupdate/js/extensionupdatecheck.es6.js#L14

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 4]
```diff
@@ -8,7 +8,107 @@ Joomla.MediaManager = Joomla.MediaManager || {};
...
+		path = path.split(':');
+        path = '/images' + path[1];
+        // Data to be saved in the storage
+		var data = "&box-left="+Joomla.MediaManager.Edit.smartcrop.cropper.boxLeft+
+                    "&box-top="+Joomla.MediaManager.Edit.smartcrop.cropper.boxTop+
+                    "&box-width="+Joomla.MediaManager.Edit.smartcrop.cropper.boxWidth+
+                    "&box-height="+Joomla.MediaManager.Edit.smartcrop.cropper.boxHeight+
+                    "&width="+width;
+
+		var xhr = new XMLHttpRequest();
```
Comment (anuragteapot) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: Same here

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 5]
```diff
@@ -8,7 +8,107 @@ Joomla.MediaManager = Joomla.MediaManager || {};
...
+		xhr.open("GET", url, true);
+		xhr.send();
+    }
+
+    // At Deactivate crop the images and save to cache.
+    function cropImages(){
+        var path = getQueryVariable('path');
+		path = path.split(':');
+        path = '/images' + path[1];
+		var xhr = new XMLHttpRequest();
```
Comment (anuragteapot) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: Same here

**Review by kasvith — APPROVED** [review]
seems fine

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 6]
```diff
@@ -78,19 +76,25 @@ public function execute($task)
 	 * 
 	 * @param   string  $imgPath  image path
 	 * 
+	 * @param   array   $widths   requested widths
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: No empty lines between `@params` lines

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 7]
```diff
@@ -46,22 +46,26 @@ public function __construct()
 	 *
 	 * index.php?option=com_media&task=adaptiveimage.setfocus&path=/images/sampledata/fruitshop/bananas_1.jpg
 	 *
-	 * @param   array   $dataFocus  Array of the values of diffrent focus point
+	 * @param   array    $dataFocus  Array of the values of diffrent focus point
+	 * 
+	 * @param   integer  $width      Width of the image
 	 *
-	 * @param   string  $imgPath    Full path for the file
+	 * @param   string   $imgPath    Full path for the file
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Same here: no empty lines between `@params` lines

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 8]
```diff
@@ -46,22 +46,26 @@ public function __construct()
...
-				"box-left"			=> $dataFocus['box-left'],
-				"box-top"			=> $dataFocus['box-top'],
-				"box-width"			=> $dataFocus['box-width'],
-				"box-height"		=> $dataFocus['box-height']
+				$width => array(
+					"box-left"			=> $dataFocus['box-left'],
+					"box-top"			=> $dataFocus['box-top'],
+					"box-width"			=> $dataFocus['box-width'],
+					"box-height"		=> $dataFocus['box-height']
+				)
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: The operators should line up.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 9]
```diff
@@ -71,15 +75,20 @@ public function setFocus($dataFocus,$imgPath)
...
 			$prevData = json_decode($prevData, true);
 
-			$prevData[$imgPath]["box-left"] 			= $dataFocus['box-left'];
-			$prevData[$imgPath]["box-top"] 				= $dataFocus['box-top'];
-			$prevData[$imgPath]["box-width"] 			= $dataFocus['box-width'];
-			$prevData[$imgPath]["box-height"] 			= $dataFocus['box-height'];
+			$prevData[$imgPath][$width]["box-left"] 			= $dataFocus['box-left'];
+			$prevData[$imgPath][$width]["box-top"] 				= $dataFocus['box-top'];
+			$prevData[$imgPath][$width]["box-width"] 			= $dataFocus['box-width'];
+			$prevData[$imgPath][$width]["box-height"] 			= $dataFocus['box-height'];
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: The operators should line up.

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 10]
```diff
@@ -42,30 +42,28 @@ public function execute($task)
 		{
 			case "setfocus" :
 				$imgPath = $this->input->getString('path');
+				$width   = $this->input->getInt('width');
 				$dataFocus = array (
 					"box-left"			=> $this->input->getInt('box-left'),
 					"box-top"			=> $this->input->getInt('box-top'),
 					"box-width"			=> $this->input->getInt('box-width'),
 					"box-height"		=> $this->input->getInt('box-height')
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: The operators should line up. Although this is no new code, follow the Scout rule: Leave the place in better shape than you found it...

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 11]
```diff
@@ -134,4 +152,49 @@ private function checkStorage($dataLocation)
...
+	 * @return  array
+	 * 
+	 * @since 4.0.0
+	 */
+	private function findMax($dataFocuses)
+	{
+		$minX=9999;
+		$minY=9999;
+		$maxWidth=0;
+		$maxHeight=0;
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Spaces around `=`. The equal signes should line up.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 12]
```diff
@@ -134,4 +152,49 @@ private function checkStorage($dataLocation)
...
+		foreach ($dataFocuses as $width => $focus)
+		{
+			if ($width!="max")
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Space around operator (also below)

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 13]
```diff
@@ -134,4 +152,49 @@ private function checkStorage($dataLocation)
...
+					$maxHeight = $focus["box-height"];
+				}
+			}
+		}
+		$maxFocus = array(
+			"box-left"			=> $minX,
+			"box-top"			=> $minY,
+			"box-width"			=> $maxWidth,
+			"box-height"		=> $maxHeight
+		);
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Line up

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 14]
```diff
@@ -8,7 +8,95 @@ Joomla.MediaManager = Joomla.MediaManager || {};
...
+            url: resolveBaseUrl() +"/administrator/index.php?option=com_media&task=adaptiveimage.cropBoxData&path="+path+"&width="+width,
+            method: 'GET',
+            onSuccess: (response) => {
+                if(response!=''){
+                    var data = JSON.parse(response);
+                    Joomla.MediaManager.Edit.smartcrop.cropper.setData({
+                    "x"	: data["box-left"],
+                    "y"	: data["box-top"],
+                    "width"	: data["box-width"],
+                    "height": data["box-height"]
```
Comment (nibra) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: Colon directly after the key

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 15]
```diff
@@ -8,7 +8,95 @@ Joomla.MediaManager = Joomla.MediaManager || {};
...
+	Joomla.MediaManager.Edit.smartcrop = {
+		Activate: function (mediaData) {
+			// Initialize
+			initSmartCrop(mediaData);
+		},
+		Deactivate: function () {
+			var width = document.getElementById("jform_requestedWidth").value;
+            saveFocusPoints(width);
+            cropImages();
+            if (!Joomla.MediaManager.Edit.smartcrop.cropper) {
```
Comment (nibra) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: Indentation...

**Review by nibra — CHANGES_REQUESTED** [review]
Be more careful regarding code style!

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 16]
```diff
@@ -8,7 +8,95 @@ Joomla.MediaManager = Joomla.MediaManager || {};
...
+    // Setting the focus area in the editor
+    var getFocusPoints = function(width){
+        Joomla.request({
+            url: resolveBaseUrl() +"/administrator/index.php?option=com_media&task=adaptiveimage.cropBoxData&path="+path+"&width="+width,
+            method: 'GET',
+            onSuccess: (response) => {
+                if(response!=''){
+                    var data = JSON.parse(response);
+                    Joomla.MediaManager.Edit.smartcrop.cropper.setData({
+                    "x"	: data["box-left"],
```
Comment (kasvith) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: indentation missing

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 17]
```diff
@@ -26,16 +114,12 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
 			minContainerHeight: image.offsetHeight,
 			crop: function (e) {
 				var left, top,
-					canvas_width, canvas_height,
-					image_width, image_height;
+					canvas_width, canvas_height;
```
Comment (kasvith) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: move up to a single line

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 18]
```diff
@@ -61,80 +145,18 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
...
-		var url = window.location.pathname.split('/');
-		if (url[1]!='administrator') {
-			return basePath+"/"+url[1];
-		}
-		return basePath;
-	}
-
-})();
\ No newline at end of file
+})();
```
Comment (kasvith) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: new line is missing

**Review by kasvith — COMMENTED** [review]
fix issues ASAP

File: media/plg_media-action_smartcrop/css/smartcrop.css (hunk 1 above)
Comment (kasvith) [inline media/plg_media-action_smartcrop/css/smartcrop.css]: Can you add a new line here?

**Review by kasvith — CHANGES_REQUESTED** [review]

File: media/plg_media-action_smartcrop/css/smartcrop.css (hunk 1 above)
Comment (geetanshu-m) [reply media/plg_media-action_smartcrop/css/smartcrop.css]: I have already added see [this](https://github.com/proman24/gsoc18_adaptive_images/blob/width_selection/media/plg_media-action_smartcrop/css/smartcrop.css).

Comment (kasvith) [reply media/plg_media-action_smartcrop/css/smartcrop.css]: github  has some issues xD

**Review by kasvith — APPROVED** [review]

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 19]
```diff
@@ -42,30 +42,28 @@ public function execute($task)
...
 				$imgPath = $this->input->getString('path');
+				$width   = $this->input->getInt('width');
 				$dataFocus = array (
-					"box-left"			=> $this->input->getInt('box-left'),
-					"box-top"			=> $this->input->getInt('box-top'),
-					"box-width"			=> $this->input->getInt('box-width'),
+					"box-left"		=> $this->input->getInt('box-left'),
+					"box-top"		=> $this->input->getInt('box-top'),
+					"box-width"		=> $this->input->getInt('box-width'),
 					"box-height"		=> $this->input->getInt('box-height')
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: Only -1- space after the longest left-value, all other aligned. NEVER use tabs but for indentation of lines.

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 20]
```diff
@@ -77,20 +75,25 @@ public function execute($task)
...
+	public function cropImage($imgPath, $widths=null)
 	{
 		$storage = new JSONFocusStore;
-		$dataFocus = json_decode($storage->getFocus($imgPath), true);
-		$width = array(240, 360, 480, 768, 940, 1024);
-		foreach ($width as $finalWidth)
+		if ( $widths==null)
+		{
+			$widths = array(240, 360, 480, 768, 940, 1024);
+
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: No empty lines before closing brace.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 21]
```diff
@@ -46,22 +46,24 @@ public function __construct()
...
 			$imgPath => array(
-				"box-left"			=> $dataFocus['box-left'],
-				"box-top"			=> $dataFocus['box-top'],
-				"box-width"			=> $dataFocus['box-width'],
-				"box-height"		=> $dataFocus['box-height']
+				$width => array(
+					"box-left"		=> $dataFocus['box-left'],
+					"box-top"		=> $dataFocus['box-top'],
+					"box-width"		=> $dataFocus['box-width'],
+					"box-height"		=> $dataFocus['box-height']
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: One-space rule also here.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 22]
```diff
@@ -71,15 +73,20 @@ public function setFocus($dataFocus,$imgPath)
...
 			$prevData = json_decode($prevData, true);
 
-			$prevData[$imgPath]["box-left"] 			= $dataFocus['box-left'];
-			$prevData[$imgPath]["box-top"] 				= $dataFocus['box-top'];
-			$prevData[$imgPath]["box-width"] 			= $dataFocus['box-width'];
-			$prevData[$imgPath]["box-height"] 			= $dataFocus['box-height'];
+			$prevData[$imgPath][$width]["box-left"] 	= $dataFocus['box-left'];
+			$prevData[$imgPath][$width]["box-top"] 		= $dataFocus['box-top'];
+			$prevData[$imgPath][$width]["box-width"] 	= $dataFocus['box-width'];
+			$prevData[$imgPath][$width]["box-height"] 	= $dataFocus['box-height'];
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: ... and here.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 23]
```diff
@@ -134,4 +149,49 @@ private function checkStorage($dataLocation)
...
+	 * 
+	 * @since 4.0.0
+	 */
+	private function findMax($dataFocuses)
+	{
+		$minX		=	9999;
+		$minY		=	9999;
+		$maxWidth	=	0;
+		$maxHeight	=	0;
+		foreach ($dataFocuses as $width => $focus)
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Always a blank line before control statements.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 24]
```diff
@@ -134,4 +149,49 @@ private function checkStorage($dataLocation)
...
+	 * @return  array
+	 * 
+	 * @since 4.0.0
+	 */
+	private function findMax($dataFocuses)
+	{
+		$minX		=	9999;
+		$minY		=	9999;
+		$maxWidth	=	0;
+		$maxHeight	=	0;
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Way too much spaces.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 25]
```diff
@@ -134,4 +149,49 @@ private function checkStorage($dataLocation)
...
+				}
+				if ($focus["box-width"] > $maxWidth)
+				{
+					$maxWidth = $focus["box-width"];
+				}
+				if ($focus["box-height"] > $maxHeight)
+				{
+					$maxHeight = $focus["box-height"];
+				}
+			}
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Better:
```php
    $minX      = min($minX, $focus["box-left"]);
    $minY      = min($minY, $focus["box-top"]);
    $maxWidth  = max($maxWidth, $focus["box-width"]);
    $maxHeight = max($maxHeight, $focus["box-height"]);
```
That's less complex.

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 26]
```diff
@@ -134,4 +149,49 @@ private function checkStorage($dataLocation)
...
+		}
+		$maxFocus = array(
+			"box-left"		=> $minX,
+			"box-top"		=> $minY,
+			"box-width"		=> $maxWidth,
+			"box-height"		=> $maxHeight
```
Comment (nibra) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: Again, too many spaces.

File: media/plg_media-action_smartcrop/js/smartcrop.js [hunk 27]
```diff
@@ -9,7 +9,95 @@ Joomla.MediaManager.Edit = Joomla.MediaManager.Edit || {};
...
+			url: resolveBaseUrl() +"/administrator/index.php?option=com_media&task=adaptiveimage.cropBoxData&path="+path+"&width="+width,
+			method: 'GET',
+			onSuccess: (response) => {
+				if(response!=''){
+					var data = JSON.parse(response);
+					Joomla.MediaManager.Edit.smartcrop.cropper.setData({
+						"x"	        : data["box-left"],
+						"y"		: data["box-top"],
+						"width"	    	: data["box-width"],
+						"height"    	: data["box-height"]
```
Comment (nibra) [inline media/plg_media-action_smartcrop/js/smartcrop.js]: Alignment and spaces

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 28]
```diff
@@ -77,20 +75,25 @@ public function execute($task)
...
 	 * 
 	 * @param   string  $imgPath  image path
+	 * @param   array   $widths   requested widths
 	 * 
 	 * @return  boolean
 	 * 
 	 * @since 4.0.0 
 	 */
-	public function cropImage($imgPath)
+	public function cropImage($imgPath, $widths=null)
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: Always 1 space around operators.

File: administrator/components/com_media/src/Controller/AdaptiveImageController.php [hunk 29]
```diff
@@ -77,20 +75,25 @@ public function execute($task)
...
 	{
 		$storage = new JSONFocusStore;
-		$dataFocus = json_decode($storage->getFocus($imgPath), true);
-		$width = array(240, 360, 480, 768, 940, 1024);
-		foreach ($width as $finalWidth)
+		if ( $widths==null)
```
Comment (nibra) [inline administrator/components/com_media/src/Controller/AdaptiveImageController.php]: Always 1 space around operators, 1 empty line before control statements.

**Review by nibra — CHANGES_REQUESTED** [review]
Still style issues :(

File: libraries/src/AdaptiveImage/JSONFocusStore.php [hunk 30]
```diff
@@ -90,13 +97,14 @@ public function setFocus($dataFocus,$imgPath)
...
-	 * @param   string  $imgPath  Image Path
+	 * @param   string   $imgPath  Image Path
+	 * @param   integer  $width    Width of the corresponding data focus point
 	 *
 	 * @return  array
 	 *
 	 * @since 4.0.0
 	 */
-	public function getFocus($imgPath)
+	public function getFocus($imgPath, $width=null)
```
Comment (kasvith) [inline libraries/src/AdaptiveImage/JSONFocusStore.php]: space around operator =

**Review by kasvith — CHANGES_REQUESTED** [review]
can you fix this

**Review by kasvith — APPROVED** [review]

**Comment by kasvith** [comment]
Nice work @proman24 👍

**Comment by geetanshu-m** [comment]
Thank you @kasvith 😄

**Pull request merged**